### PR TITLE
Added support for STM32F412Zx MCU

### PR DESCRIPTION
**Pull request by Toritos01**
This PR fixes/implements generic support for the STM32F412Zx variant, using the steps at this link: https://github.com/stm32duino/Arduino_Core_STM32/wiki/Add-a-new-variant-%28board%29

The motivation for stm32duino to have support for more variants, and to allow me to use this repo with a board that I personally have.

### CODE REVIEW

File: variants/STM32F4xx/F412Z(E-G)(J-T)/generic_clock.c [hunk 1]
```diff
@@ -10,19 +10,55 @@
...
-  */
+ * @brief  System Clock Configuration
+ * @param  None
+ * @retval None
+ */
 WEAK void SystemClock_Config(void)
 {
   /* SystemClock_Config can be generated by STM32CubeMX */
-#warning "SystemClock_Config() is empty. Default clock at reset is used."
+  // #warning "SystemClock_Config() is empty. Default clock at reset is used."
```
Comment (fpistm) [inline variants/STM32F4xx/F412Z(E-G)(J-T)/generic_clock.c:25]:
```suggestion
```

**Review by fpistm — CHANGES_REQUESTED** [review]
Some minor changes to be done.
Is USB clock configured to 48MHz?

**Comment by Toritos01** [comment]
> Some minor changes to be done. Is USB clock configured to 48MHz?

I tried enabling UST_OTG_FS on STM32CubeMx, but doing so breaks the clock configuration

File: variants/STM32F4xx/F412Z(E-G)(J-T)/generic_clock.c [hunk 2]
```diff
@@ -22,7 +22,42 @@
...
+   */
+  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
+  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
+  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
+  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV2;
+  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;
+
+  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_3) != HAL_OK) {
+    Error_Handler();
+  }
```
Comment (fpistm) [inline variants/STM32F4xx/F412Z(E-G)(J-T)/generic_clock.c:60]: Could you try this:

```suggestion
  RCC_OscInitTypeDef RCC_OscInitStruct = {};
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
  RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {};

  /** Configure the main internal regulator output voltage
  */
  __HAL_RCC_PWR_CLK_ENABLE();
  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);

  /** Initializes the RCC Oscillators according to the specified parameters
  * in the RCC_OscInitTypeDef structure.
  */
  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSI;
  RCC_OscInitStruct.HSIState = RCC_HSI_ON;
  RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSI;
  RCC_OscInitStruct.PLL.PLLM = 8;
  RCC_OscInitStruct.PLL.PLLN = 96;
  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
  RCC_OscInitStruct.PLL.PLLQ = 4;
  RCC_OscInitStruct.PLL.PLLR = 2;
  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
    Error_Handler();
  }

  /** Initializes the CPU, AHB and APB buses clocks
  */
  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK
                                | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV2;
  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;

  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_3) != HAL_OK) {
    Error_Handler();
  }

  /** Initializes the peripherals clock
  */
  PeriphClkInitStruct.PeriphClockSelection = RCC_PERIPHCLK_CLK48;
  PeriphClkInitStruct.Clk48ClockSelection = RCC_CLK48CLKSOURCE_PLLQ;
  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK) {
    Error_Handler();
  }
```

File: variants/STM32F4xx/F412Z(E-G)(J-T)/generic_clock.c [hunk 3]
```diff
@@ -22,7 +22,42 @@
 WEAK void SystemClock_Config(void)
 {
   /* SystemClock_Config can be generated by STM32CubeMX */
```
Comment (fpistm) [inline variants/STM32F4xx/F412Z(E-G)(J-T)/generic_clock.c:24]: This line can be deleted
```suggestion
```

**Review by fpistm — CHANGES_REQUESTED** [review]
Thanks for the update.
One line to remove and a proposal for system core clock config to have USB clokk at 48Mhz.
If you can test it and give your feedback, I could not test as I do not have a STM32F412Zx board.

**Review by fpistm — APPROVED** [review]

**Comment by Toritos01** [comment]
Sorry for the delay with testing this, didn't have access to my STM32F4 for a few days. I tested out the proposed fixes, but when I try running the CheckVariant.ino program, my LED does not blink and nothing shows up the serial monitor. Any recommendations on additional steps to take for this to work?

**Comment by fpistm** [comment]
Hi @Toritos01 
No worry.
The main changes was to configure clock for the USB:

```patch
+  RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {};
 
   /** Configure the main internal regulator output voltage
  */
   __HAL_RCC_PWR_CLK_ENABLE();
   __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);
 
   /** Initializes the RCC Oscillators according to the specified parameters
  * in the RCC_OscInitTypeDef structure.
  */
   RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSI;
   RCC_OscInitStruct.HSIState = RCC_HSI_ON;
   RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
   RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
   RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSI;
   RCC_OscInitStruct.PLL.PLLM = 8;
-  RCC_OscInitStruct.PLL.PLLN = 100;
+  RCC_OscInitStruct.PLL.PLLN = 96;
   RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
-  RCC_OscInitStruct.PLL.PLLQ = 2;
+  RCC_OscInitStruct.PLL.PLLQ = 4;
   RCC_OscInitStruct.PLL.PLLR = 2;
   if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
     Error_Handler();
   }
 
   /** Initializes the CPU, AHB and APB buses clocks
-   */
-  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
+  */
+  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK
+                                | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
   RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
   RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
   RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV2;
@@ -58,6 +59,14 @@ WEAK void SystemClock_Config(void)
   if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_3) != HAL_OK) {
     Error_Handler();
   }
+
+  /** Initializes the peripherals clock
+  */
+  PeriphClkInitStruct.PeriphClockSelection = RCC_PERIPHCLK_CLK48;
+  PeriphClkInitStruct.Clk48ClockSelection = RCC_CLK48CLKSOURCE_PLLQ;
+  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK) {
+    Error_Handler();
+  }
```

It was generated thanks CubeMx and should be fine.
Main changes is we change the clock freq from 100 to 96 MHz in order to have a 48MHz for USB.
That's strange it does not works. I guess it fails in the `HAL_RCCEx_PeriphCLKConfig` but don't know why. You could try to debug with Arduino IDE 2.1.0 by enabling the -Og and -g in the menu. Then you should be able to debug.

**Comment by fpistm** [comment]
Fixes #1899

**Pull request merged**